cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java

```diff
@@ -71,6 +80,184 @@ public Schema getSchema() {
+    Instant instant = Instant.ofEpochSecond(tsInSeconds, unit.toNanos(fraction));
+    return ZonedDateTime.ofInstant(instant, zoneId);
+  }
+
+  /**
+   * Validates and returns the underlying {@link LogicalType} for the given field.
+   *
+   * @param fieldName name of the field
+   * @param allowedTypes acceptable logical types
+   * @return the underlying {@link LogicalType} for the field
+   * @throws IllegalArgumentException if the field name is not found in the map of fields or the field provided
+   *                                  is not an acceptable type
+   */
+  private LogicalType validateAndGetLogicalType(String fieldName, Set<LogicalType> allowedTypes) {
+    Schema.Field field = schema.getField(fieldName);
+    if (field == null) {
+      throw new IllegalArgumentException(String.format("Field %s does not exists in the schema", fieldName));
+    }
+    return validateAndGetLogicalType(field, allowedTypes);
+  }
+
+  /**
+   * Validates and returns the underlying {@link LogicalType} of the given {@link Schema.Field}.
+   *
+   * @param field field with logical type
+   * @param allowedTypes acceptable logical types
+   * @return the underlying {@link LogicalType} for the field
+   * @throws IllegalArgumentException if the field provided does not have a {@link LogicalType} that is
+   *                                  one of the acceptable types
+   */
+  @Nullable
+  private static LogicalType validateAndGetLogicalType(Schema.Field field, Set<LogicalType> allowedTypes) {
+    if (field == null) {
+      return null;
+    }
+
+    String fieldName = field.getName();
+    LogicalType logicalType = getLogicalType(field.getSchema(), allowedTypes);
+
+    if (logicalType == null) {
+      throw new IllegalArgumentException(String.format("Field %s does not have a logical type.", fieldName));
+    }
+
+    if (!allowedTypes.contains(logicalType)) {
+      throw new IllegalArgumentException(String.format("Field %s must be of logical type %s, instead it is of type %s",
+                                                       fieldName, allowedTypes, logicalType));
+    }
+    return logicalType;
+  }
+
+  /**
+   * Gets the {@link LogicalType} defined in the given {@link Schema}.
+   * If the given {@link Schema} is a {@link Schema.Type#UNION}, this method will find
+   * the {@link LogicalType} recursively and return the first {@link LogicalType} it encountered.
+   *
+   * @param schema the {@link Schema} for finding the {@link LogicalType}
+   * @param allowedTypes acceptable logical types
+   * @return the {@link LogicalType} defined in the given schema or {@code null} if no {@link LogicalType} was found
+   */
+  @Nullable
+  private static LogicalType getLogicalType(Schema schema, Set<LogicalType> allowedTypes) {
+    LogicalType logicalType = schema.getLogicalType();
+
+    if (logicalType != null && allowedTypes.contains(logicalType)) {
```

Comment:
We need `allowedTypes.contains(logicalType)` for the union schema which has multiple logical types, for example:
```java
Schema schema = Schema.recordOf("x", Schema.Field.of("x", Schema.unionOf(
      Schema.of(Schema.LogicalType.DATE),
      Schema.nullableOf(Schema.of(Schema.LogicalType.TIME_MILLIS)))));
```

Reply:
Which one to return if both are allowed? The first one?

Reply:
In that case first matching allowed logical type will be used. For example, if the schema is:
```java
 Schema schema = Schema.recordOf("x", Schema.Field.of("x", Schema.unionOf(
      Schema.unionOf(Schema.nullableOf(Schema.of(Schema.LogicalType.TIMESTAMP_MILLIS)),
                     Schema.of(Schema.Type.STRING)),
      Schema.of(Schema.Type.NULL),
      Schema.of(Schema.Type.INT),
      Schema.of(Schema.Type.LONG),
      Schema.of(Schema.LogicalType.DATE),
      Schema.nullableOf(Schema.of(Schema.LogicalType.TIME_MILLIS)),
      Schema.nullableOf(Schema.of(Schema.LogicalType.TIME_MICROS)))));
```

```java
// Sets Time in millis
StructuredRecord.builder(schema).setTime("x", time);
```